src/main/java/com/es/phoneshop/model/dao/GenericDaoImpl.java

```diff
@@ -0,0 +1,46 @@
+package com.es.phoneshop.model.dao;
+
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.model.product.ProductNotFoundException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public abstract class GenericDaoImpl<T extends AbstractDao> implements GenericDao<T> {
```

Comment:
I got your concept here and generally it's ok, but you should change class namings. 1) AbstractDao - this interface represents common methods of beans dao is processing, but it's not dao in fact. So it's should be named like Bean or DaoBean or even DaoEntity, think you got my thoughs. 2) Usually classes named Generic... are abstract classes that share some common code. Better to rename like GenericDao->Dao GenericDaoImpl->GenericDao with leaving interface and abstract class as is, you can compare to servlet hierarchy. In general the way you constructed it is fine.